presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -605,13 +610,20 @@ private static boolean typesMatchForInsert(Iterable<Type> tableTypes, Iterable<T
                 if (!tableType.getTypeSignature().getBase().equals(queryType.getTypeSignature().getBase())) {
                     return false;
                 }
-                if (!typesMatchForInsert(tableType.getTypeParameters(), queryType.getTypeParameters())) {
+                if (!typesMatchForInsert(tableType.getTypeParameters(), queryType.getTypeParameters(), true)) {
                     return false;
                 }
             }
             else {
-                if (!Objects.equals(tableType, queryType) && !isTypeOnlyCoercion(queryType.getTypeSignature(), tableType.getTypeSignature())) {
-                    return false;
+                if (structuredType) {
```

Comment:
Why do we have 2 different paths for primitive and structured types? Can't it be just one single path? I think it should be possible to insert `ARRAY[INTEGER]` to `ARRAY[DOUBLE]` column. As i know only `MAP` cast operator is not specified.